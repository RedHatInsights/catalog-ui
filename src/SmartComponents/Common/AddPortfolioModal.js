import React, { Component } from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { withRouter } from 'react-router-dom';
import propTypes from 'prop-types';
import { Main, PageHeader, PageHeaderTitle } from '@red-hat-insights/insights-frontend-components';
import { addPortfolioWithItem, fetchPortfolios } from '../../Store/Actions/PortfolioActions';
import { PortfolioStore } from '../../Store/Reducers/PortfolioStore';
import { consoleLog } from '../../Helpers/Shared/Helper';
import DataDrivenForm from './DataDrivenForm'
import { addAlert, removeAlert } from '../../Store/Actions/AlertActions';

const schema = {
    type: 'object',
    properties: {
        name: { title: 'New Portfolio Name', type: 'string' },
        description: { title: 'Description', type: 'string' }
    },
    required: [ 'name', 'description' ]
};

class AddPortfolioModal extends Component {
    componentDidMount() {
    };

  onSubmit = data => {
      this.props.addPortfolioWithItem(data, this.props.itemdata)
      .then(() => { this.props.addAlert({
          variant: 'success',
          title: 'Success adding portfolio',
          description: 'The portfolio was added successfully.'
      });
      this.props.fetchPortfolios();
      })
      .catch(() => this.props.addAlert({
          variant: 'danger',
          title: 'Failed adding portfolio',
          description: 'The portfolio was not added successfuly.'
      }));
      this.props.closeModal();
  }

  onCancel = () => {
      this.props.addAlert({
          variant: 'warning',
          title: 'Adding portfolio',
          description: 'Adding portfolio was cancelled by the user.'
      });
      this.props.closeModal();
  }

  render() {
      consoleLog('Adding a New Portfolio');
      return (
          <Main title={ 'Add Portfolio' }>
              <div className="pf-l-stack">
                  <div className="pf-l-stack__item pf-m-secondary ">
                      <PageHeader>
                          <PageHeaderTitle title= 'Create Portfolio and Add Product' />
                      </PageHeader>
                  </div>
                  <div className="pf-l-stack">
                      <DataDrivenForm
                        schemaType="mozilla"
                        schema={ schema }
                        onSubmit={ this.onSubmit }
                        onCancel={ this.onCancel } />
                  </div>
              </div>
          </Main>
      );
  }
}

const mapStateToProps = state => ({ isLoading: state.PortfolioStore.isLoading });

const mapDispatchToProps = dispatch => bindActionCreators({
    addAlert,
    removeAlert,
    addPortfolioWithItem,
    fetchPortfolios
}, dispatch);

AddPortfolioModal.propTypes = {
    isLoading: propTypes.bool,
    history: propTypes.object,
    addAlert: propTypes.func,
    addPortfolioWithItem: propTypes.func,
    fetchPortfolios: propTypes.func,
    closeModal: propTypes.func,
    itemdata: propTypes.object
};

export default withRouter(connect(mapStateToProps, mapDispatchToProps)(AddPortfolioModal));
