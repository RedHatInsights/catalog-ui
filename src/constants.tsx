export class Constants {
  static readonly INSIGHTS_DEPLOYMENT_MODE = 'insights';
  static readonly STANDALONE_DEPLOYMENT_MODE = 'standalone';
}
