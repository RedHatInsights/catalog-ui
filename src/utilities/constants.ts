export const CATALOG_API_BASE = `${process.env.BASE_PATH}/catalog/v1.2`;
export const SOURCES_API_BASE = `${process.env.BASE_PATH}/sources/v1.0`;
export const APPROVAL_API_BASE = `${process.env.BASE_PATH}/approval/v1.2`;
export const TOPOLOGICAL_INVENTORY_API_BASE = `${process.env.BASE_PATH}/topological-inventory/v1.0`;
export const RBAC_API_BASE = `${process.env.BASE_PATH}/rbac/v1`;
export const permissionValues = ['order', 'read', 'update'];

export const permissionOptions = [
  {
    value: 'order,read,update',
    label: 'Can order/edit'
  },
  {
    value: 'order,read',
    label: 'Can order/view'
  }
];

export const PORTFOLIO_RESOURCE_TYPE = 'Portfolio';
export const PORTFOLIO_ITEM_RESOURCE_TYPE = 'PortfolioItem';
export const INVENTORY_RESOURCE_TYPE = 'ServiceInventory';
export const DEFAULT_MAX_LENGTH = 64;

export const APP_NAME = {
  Portfolio: 'catalog',
  PortfolioItem: 'catalog',
  ServiceInventory: 'topology'
};

export const BEFORE_TYPE = 'before';
export const AFTER_TYPE = 'after';
export const MAX_APPROVAL_REQ_RETRIES = 5;
