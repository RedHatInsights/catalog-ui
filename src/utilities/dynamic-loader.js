/* eslint-disable camelcase */
import React from 'react';

const useDynamicScript = (args) => {
  const scriptElement = React.useRef(null);
  const [ready, setReady] = React.useState(false);
  const [failed, setFailed] = React.useState(false);
  const getEntry = async (url) => {
    let manifest = await fetch(url).then((d) => d.json());
    manifest = manifest.approval.entry[0];
    const result = manifest;

    const element = document.createElement('script');

    element.src = result;
    element.type = 'text/javascript';
    element.async = true;

    setReady(false);
    setFailed(false);

    element.onload = () => {
      console.log(`Dynamic Script Loaded: ${result}`);
      scriptElement.current = element;
      setReady(true);
    };

    element.onerror = () => {
      console.error(`Dynamic Script Error: ${result}`);
      setReady(false);
      setFailed(true);
    };

    document.head.appendChild(element);
  };

  React.useEffect(() => {
    if (!args.url) {
      return;
    }

    getEntry(args.url);

    return () => {
      if (scriptElement.current) {
        console.log(`Dynamic Script Removed: ${args.url}`);
        document.head.removeChild(scriptElement.current);
      }
    };
  }, [args.url]);

  return {
    ready,
    failed
  };
};

const loadComponent = (scope, module) => async () => {
  // Initializes the share scope. This fills it with known provided modules from this build and all remotes
  // eslint-disable-next-line no-undef
  await __webpack_init_sharing__('default');

  const container = window[scope]; // or get the container somewhere else
  // Initialize the container, it may provide shared modules
  // eslint-disable-next-line no-undef
  await container.init(__webpack_share_scopes__.default);
  const factory = await window[scope].get(module);
  const Module = factory();
  return Module;
};

export { useDynamicScript, loadComponent };
