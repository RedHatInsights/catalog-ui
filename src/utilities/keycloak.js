import Keycloak from 'keycloak-js';

const _keycloak = new Keycloak('./keycloak.json');

/**
 * Initializes Keycloak instance and calls the provided callback function if successfully authenticated.
 *
 * @param onAuthenticatedCallback
 */
const initKeycloak = (onAuthenticatedCallback) => {
  _keycloak
    .init({
      onLoad: 'check-sso',
      silentCheckSsoRedirectUri:
        window.location.origin + '/silent-check-sso.html',
      pkceMethod: 'S256'
    })
    .then((authenticated) => {
      if (!authenticated) {
        console.log('user is not authenticated..!');
      }

      onAuthenticatedCallback();
    })
    .catch(console.error);
};

const doLogin = _keycloak.login;

const doLogout = _keycloak.logout;

const getToken = () => _keycloak.token;

const isLoggedIn = () => !!_keycloak.token;

const updateToken = (successCallback) =>
  _keycloak
    .updateToken(5)
    .then(successCallback)
    .catch(doLogin);

const getUsername = () => _keycloak.tokenParsed?.preferred_username;

const hasRole = (roles) => roles.some((role) => _keycloak.hasRealmRole(role));

const CatalogAuth = {
  initKeycloak,
  doLogin,
  doLogout,
  isLoggedIn,
  getToken,
  updateToken,
  getUsername,
  hasRole
};

export default CatalogAuth;
