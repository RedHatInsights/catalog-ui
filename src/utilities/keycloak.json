{
  "realm": "aap",
  "auth-server-url": "http://keycloak.k8s.local/auth",
  "ssl-required": "external",
  "resource": "catalog",
  "verify-token-audience": true,
  "credentials": {
    "secret": "SOMESECRETVALUE"
  },
  "use-resource-role-mappings": true,
  "confidential-port": 0,
  "policy-enforcer": {}
}
