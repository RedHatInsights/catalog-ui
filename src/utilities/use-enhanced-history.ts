import { useHistory } from 'react-router-dom';
import { History, Path } from 'history';
import { AnyObject } from '@data-driven-forms/react-form-renderer';

export type EnhancedHistoryPushtarget =
  | Path
  | { pathname: string; hash?: string; search?: string; state?: AnyObject };

const removeSearchQuery = (target: EnhancedHistoryPushtarget) => {
  if (typeof target === 'string') {
    return target.split('?')[0];
  }

  if (typeof target === 'object') {
    return {
      ...target,
      search: undefined
    };
  }

  return target;
};

/**
 * Wrapper around useHistory hook which adds aditional settings to standard history methods
 * @param {Boolean} removeSearch if true, using history navigation methods will remove search string from path
 * @param {Boolean} keepHash if true, using history navigation methods will not remove hash from URL
 */
const useEnhancedHistory = ({
  removeSearch,
  keepHash
}: {
  removeSearch?: boolean;
  keepHash?: boolean;
} = {}): History<History.LocationState> => {
  const history = useHistory();
  return {
    ...history,
    push: (target: EnhancedHistoryPushtarget) => {
      let internalTarget = target;
      if (keepHash && history.location.hash.length > 0) {
        internalTarget =
          typeof internalTarget === 'object'
            ? { ...internalTarget, hash: history.location.hash }
            : `${internalTarget}${history.location.hash}`;
      }

      if (removeSearch) {
        return history.push(removeSearchQuery(internalTarget));
      }

      return history.push(internalTarget);
    }
  };
};

export default useEnhancedHistory;
