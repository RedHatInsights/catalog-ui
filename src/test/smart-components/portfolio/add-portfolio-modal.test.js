import React from 'react';
import thunk from 'redux-thunk';
import { shallow, mount } from 'enzyme';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store' ;
import { Modal } from '@patternfly/react-core';
import { shallowToJson } from 'enzyme-to-json';
import { MemoryRouter, Route } from 'react-router-dom';
import promiseMiddleware from 'redux-promise-middleware';
import { componentTypes } from '@data-driven-forms/react-form-renderer';
import { notificationsMiddleware } from '@redhat-cloud-services/frontend-components-notifications/';

import { APPROVAL_API_BASE, CATALOG_API_BASE } from '../../../utilities/constants';
import FormRenderer from '../../../smart-components/common/form-renderer';
import AddPortfolioModal from '../../../smart-components/portfolio/add-portfolio-modal';

describe('<AddPortfolioModal />', () => {
  let initialProps;
  let initialState;
  const middlewares = [ thunk, promiseMiddleware(), notificationsMiddleware() ];
  let mockStore;
  const ComponentWrapper = ({ store, children, portfolioId }) => (
    <Provider store={ store }>
      <MemoryRouter initialEntries={ [ `portfolios/${portfolioId}` ] }>
        { children }
      </MemoryRouter>
    </Provider>
  );

  beforeEach(() => {
    initialProps = {
      fetchPortfolios: jest.fn()
    };
    initialState = {
      portfolioReducer: {
        portfolios: { data: [{
          id: '123',
          name: 'Portfolio'
        }]}
      }
    };
    mockStore = configureStore(middlewares);
  });

  it('should render correctly', () => {
    const store = mockStore({});
    const wrapper = shallow(<ComponentWrapper store={ store }><AddPortfolioModal { ...initialProps } /></ComponentWrapper>).dive();

    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });

  it('should create edit variant of portfolio modal', done => {
    const store = mockStore(initialState);

    const expectedSchema = {
      fields: [{
        component: componentTypes.TEXT_FIELD,
        isRequired: true,
        label: 'Portfolio Name',
        name: 'name',
        validate: [ expect.any(Function) ]
      }, {
        component: componentTypes.TEXTAREA,
        label: 'Description',
        name: 'description'
      }]
    };

    const wrapper = mount(
      <ComponentWrapper store={ store } portfolioId="123">
        <Route path="portfolios/:id?" render={ () => <AddPortfolioModal { ...initialProps } match={ { params: { id: '123' }} } /> }/>
      </ComponentWrapper>
    );

    setImmediate(() => {
      const modal = wrapper.find(Modal);
      const form = wrapper.find(FormRenderer);
      expect(modal.props().title).toEqual('Edit portfolio');
      expect(form.props().schema).toEqual(expectedSchema);
      done();
    });
  });

  it('should create edit variant of portfolio modal and call updatePortfolio on submit', () => {
    const store = mockStore(initialState);

    apiClientMock.patch(`${CATALOG_API_BASE}/portfolios/123`, ((req, res) => {
      expect(JSON.parse(req.body())).toEqual({ id: '123', name: 'Portfolio', workflow_ref: null });
      return res.body(200);
    }));

    const wrapper = mount(
      <ComponentWrapper store={ store } portfolioId="123">
        <Route
          path="portfolios/:id?"
          render={ () => <AddPortfolioModal { ...initialProps } match={ { params: { id: '123' }} }/> }
        />
      </ComponentWrapper>
    );

    const button = wrapper.find('button').last();
    button.simulate('click');
  });
});
