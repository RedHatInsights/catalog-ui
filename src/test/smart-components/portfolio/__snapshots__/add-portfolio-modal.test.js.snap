// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AddPortfolioModal /> should render correctly 1`] = `
<ContextProvider
  value={
    Object {
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "subscription": Subscription {
        "handleChangeWrapper": [Function],
        "listeners": Object {
          "clear": [Function],
          "get": [Function],
          "notify": [Function],
          "subscribe": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "unsubscribe": [Function],
      },
    }
  }
>
  <MemoryRouter
    initialEntries={
      Array [
        "portfolios/undefined",
      ]
    }
  >
    <withRouter(Connect(AddPortfolioModal))
      fetchPortfolios={[MockFunction]}
    />
  </MemoryRouter>
</ContextProvider>
`;
